# 二元函数

二元运算是指作用于两个对象（例如数组）进行的运算：
给定集合 `A` ,二元函数 `F` : $A\times A$ → $A$ 称为集合 `A` 上的二元运算。
需要注意的是，二元运算的运算结果跟两个输入值必须是同种东西。例如，整数的加法是二元运算，因为整数相加后仍然是整数。

In [87]:
import numpy as np

## 简单操作
将两个数组各元素进行对应运算

在正式开始前，我们先生成数组 `x` 和 `y` 用作演示练习使用。

In [88]:
x = np.array([1,2,3,4])
x

array([1, 2, 3, 4])

In [89]:
y = np.array([-1.2,1.4,1.5,1.7])
y

array([-1.2,  1.4,  1.5,  1.7])

### 基础运算
对于两个数组个元素对应进行加减乘运算。

In [90]:
x+y

array([-0.2,  3.4,  4.5,  5.7])

In [91]:
x-y

array([2.2, 0.6, 1.5, 2.3])

In [92]:
x*y

array([-1.2,  2.8,  4.5,  6.8])

### 最值计算
对两个数组间进行元素级最值计算

#### 1.最大值计算
我们可以使用 `np.fmax` 和 `np.maximum` 函数对两个数组 `x` 和 `y` 进行元素级最大值计算

In [93]:
a1=np.fmax(x,y)
a1

array([1., 2., 3., 4.])

In [94]:
a2=np.maximum(x,y)
a2

array([1., 2., 3., 4.])

#### 2.最小值计算
我们可以使用 `np.fmin` 和 `np.minimum` 函数对两个数组 `x` 和 `y` 进行元素级最大值计算

In [95]:
a1=np.fmin(x,y)
a1

array([-1.2,  1.4,  1.5,  1.7])

In [96]:
a2=np.minimum(x,y)
a2

array([-1.2,  1.4,  1.5,  1.7])

### 元素及模运算
我们可以使用 `np.mod` 函数对两个数组元素进行模运算。

In [97]:
np.mod(x,y)

array([-0.2,  0.6,  0. ,  0.6])

除此之外，我们可以使用 `np.copysign` 函数，将数组 `y` 中各元素的符号赋值给数组 `x` 的对应元素。

In [98]:
np.copysign(x,y)

array([-1.,  2.,  3.,  4.])

### 数组算数比较
我们可以对两个数组进行算数比较，产生布尔型数组。
这里我们使用的函数符号有：
`＜` （小于），`＞` （大于）， `＞=` （大于等于）， `＜=` （小于等于）， `==` （等于），`！=` （不等于）。
需要注意的是，我们常用 `=` 表示赋值功能，因此在判定数组是否相等中，使用 `==` 符号。

下面用一个例子作为示范，使用 `==` 对数组 `x` ， `y` 元素级进行比较是否相等，相等输出结果 `True` ，若元素对应不相等，则输出结果 `False`。

In [99]:
x==y

array([False, False, False, False])

由此我们可以知道，对于该两个数组 `x` 和 `y` ，每一个对应元素值都不相等。

## 按元素位操作（Elementwise bit operations）

### 按位与（&）

按位与（`&`）运算符是，在两个操作数对应的二进制位都为 `1` 时，该位的结果值才为 `1`，将运算符应用于每对位，然后按位构造结果。
| a    	| b                     	| a AND b|
|:---:	|:---:	|:---:|
| 0  	| 0 	| 0 |
| 0 	| 1     | 0 |
| 1 	| 0     | 0 |
| 1     | 1     |1  |  

性质：任何数组 `x` 与 `0` 进行按位与运算都会得到数字 `0`。

`bitwise_and` 对两个数组按元素计算按位与（bit-wise AND）。
让我们通过一个例子实践一下，数字 `9` （十进制）可以用二进制表示为 `0000 1001` 。数字 `23` 同样可以用二进制表示为 `0001 0111`。则 `9` 与`23` 的按位与（bit-wise AND）就是 `0000 0001` ，或者说1。

In [100]:
np.bitwise_and(9,23)

1

更进一步，`21` (base 10) = `0001 0101`。
则 `23` 与`21` 做按位与（bit-wise AND）计算 `0001 0101`, 或者说21.

In [101]:
np.bitwise_and(21,23)

21

In [102]:
np.binary_repr(21) #使用二进制表示

'10101'

接下来，我们可以使用 `bitwise_and` 对数组（array）进行按位与运算，下面是一些例子。

In [103]:
np.bitwise_and([9,21],23)

array([ 1, 21], dtype=int32)

In [104]:
np.bitwise_and([9,21],[23,14])

array([1, 4])

In [105]:
np.bitwise_and(np.array([2,5,255]), np.array([3,14,16]))

array([ 2,  4, 16])

In [106]:
np.bitwise_and([True, False], [False, False])

array([False, False])

在给定数组的情况下，该函数操作可以简化为 `&` ：

In [107]:
x1 = np.array([2,5,255])
x2 = np.array([3,14,16])
x1 & x2

array([ 2,  4, 16])

除了按位与（`&`）外，位操作函数还有：
`bitwise_or` 或
`bitwise_xor` 异或

### 按位或（|）


| a    	| b                     	| a AND b|
|:---:	|:---:	|:---:|
| 0  	| 0 	| 0 |
| 0 	| 1     | 1 |
| 1 	| 0     | 1 |
| 1     | 1     | 1 |  


`bitwise_or` 对两个数组按元素计算按位或（bit-wise OR）。
让我们通过一个例子实践一下，数字 `9` （十进制）可以用二进制表示为 `0000 1001` 。数字 `23` 同样可以用二进制表示为 `0001 0111`。则 `9` 与`23` 的按位或（bit-wise OR）就是 `0001 1111` ，或者说 `31`。

In [108]:
np.bitwise_or(9,23)

31

接下来，我们可以使用 `bitwise_or` 对数组（array）进行按位与运算，下面是更多的一些例子。

In [109]:
np.bitwise_or([9,21],23)

array([31, 23], dtype=int32)

In [110]:
np.bitwise_or([9,21],[23,14])

array([31, 31])

In [111]:
np.bitwise_or(np.array([2,5,255]), np.array([3,14,16]))

array([  3,  15, 255])

In [112]:
np.bitwise_or([True, False], [False, False])

array([ True, False])

在给定数组的情况下，该函数操作可以简化为 `|` ：

In [113]:
x1 = np.array([2,5,255])
x2 = np.array([3,14,16])
x1 | x2

array([  3,  15, 255])

### 异或（^）
| a    	| b                     	| a AND b|
|:---:	|:---:	|:---:|
| 0  	| 0 	| 0 |
| 0 	| 1     | 1 |
| 1 	| 0     | 1 |
| 1     | 1     | 0 |  

让我们通过一个例子实践一下，数字 `9` （十进制）可以用二进制表示为 `0000 1001` 。数字 `23` 同样可以用二进制表示为 `0001 0111`。则 `9` 与`23` 的按位或（bit-wise XOR）就是 `0001 1110` ，或者说 `30`。

In [114]:
np.bitwise_xor(9,23)

30

In [115]:
np.bitwise_xor([9,21], [23,14])

array([30, 27])

In [116]:
np.bitwise_xor([True, True], [False, True])

array([ True, False])

在给定数组的情况下，该函数操作可以简化为 `^` ：

In [117]:
x1 = np.array([2,5,255])
x2 = np.array([3,14,16])
x1 ^ x2

array([  1,  11, 239])

### 按位取否（bit-wise NOT）
`invert()` 函数计算输入数组中整数的二进制按位取否的结果。

下面举例来看 `invert()` 函数的作用。
`18` (base 10) = `0001 0010` (base 2), 那么对 `18` 按位取否（bit-wise NOT）得到：

In [118]:
x = np.invert(np.array(18, dtype=np.uint8))
x

237

In [119]:
np.binary_repr(x, width=8)

'11101101'

需要注意的是，输出的结果取决于 `bit_width` 。
接着上面的例子，当 `dtype` 取 `16` 时，结果会变得不同：

In [120]:
y = np.invert(np.array(18, dtype=np.uint16))
y

65517

In [121]:
np.binary_repr(y, width=16)

'1111111111101101'

布尔型也可以使用：

In [122]:
np.invert(np.array([True, False]))

array([False,  True])

在给定数组的情况下，该函数操作可以简化为 `~` ：

In [123]:
x1 = np.array([2,5,8], dtype=np.uint8)
~x1

array([253, 250, 247], dtype=uint8)

In [124]:
x2 = np.array([True, False])
~x2

array([False,  True])

### 移位
将输入数组的整数按二进制的位向左/向右移位。

#### 向左移位
`letf_shift` 将输入数组的整数按二进制的位向左移位。
具体的操作是，通过在 `x1` 的右边附加 `x2 ` 个 `0` ，从而向左移位。
由于数字是二进制格式表示，所以该运算等价于 `x1×2^x2` 。

下面通过举例来看 `left_shift` 具体作用。
`5` (base 10) = `0101` (base 2)：

In [125]:
np.left_shift(5, 2)

20

In [126]:
np.binary_repr(20)

'10100'

In [127]:
np.left_shift(5, [1,2,3])

array([10, 20, 40], dtype=int32)

需要注意的是，第二个参数的 `dtype` 可能会改变结果的 `dtype`:

例如，`255` (base 10) = `1111 1111` (base 2)，`254` (base 10) = `1111 1110` (base 2)。
我们希望将 `255` 向左移动 `1` 位得到  `254` ：

In [128]:
a = np.left_shift(np.uint8(255), 1) 
print(a, type(a))

510 <class 'numpy.intc'>


得到结果510，并不是我们期望的结果，现在我们限定第二个参数的 `dtype`：

In [129]:
b = np.left_shift(np.uint8(255), np.uint8(1))
print(b, type(b))

254 <class 'numpy.uint8'>


以上便得到了我们所希望移位后得到的结果。

该函数操作可以简化为 `<<`：

In [130]:
x1 = 5
x2 = np.array([1, 2, 3])
x1 << x2

array([10, 20, 40], dtype=int32)

#### 向右移位
`right_shift` 将输入数组的整数按二进制的位向左移位。
具体的操作是，将 `x1` 的向右按位移动 `x2 ` ，从而向右移位。

下面通过举例来看 `right_shift` 具体作用。
`10` (base 10) = `1010` (base 2)：

In [131]:
np.right_shift(10, 1)

5

In [132]:
np.binary_repr(5)

'101'

In [133]:
np.right_shift(10, [1,2,3])

array([5, 2, 1], dtype=int32)

该函数操作可以简化为 `>>`：

In [134]:
x1 = 10
x2 = np.array([1,2,3])
x1 >> x2

array([5, 2, 1], dtype=int32)

## 按位压缩存储（Bit packing）

`packbits()` 函数，将二进制数值数组的元素打包成一个 `unit8` 数组中的位。
具体操作是通过在数组中的整数末尾插入 `0` ，使输出结果被填充为完整字节。

In [135]:
a = np.array([[[1,0,1],
               [0,1,0]],
              [[1,1,0],
               [0,0,1]]])
b = np.packbits(a, axis=-1)
b

array([[[160],
        [ 64]],

       [[192],
        [ 32]]], dtype=uint8)

`160` (base 10) = `1010 0000` ；
`64` (base 10) = `0100 0000` ；
`192` (base 10) = `1100 0000` ；
`32` (base 10) = `0010 0000` 。

`unpackbits()` 函数，将 `uint8` 数组的元素解压缩到二进制值的输出数组中。

每个元素都表示一个应解压缩到二进制值输出数组中的位字段。输出数组的形状可以是一维的（如果 `axis=None` ），也可以是与输入数组相同的形状，并沿着指定的轴进行解压。

In [136]:
a = np.array([[2], [7], [23]], dtype=np.uint8)
b = np.unpackbits(a, axis=1)
b

array([[0, 0, 0, 0, 0, 0, 1, 0],
       [0, 0, 0, 0, 0, 1, 1, 1],
       [0, 0, 0, 1, 0, 1, 1, 1]], dtype=uint8)

In [137]:
c = np.unpackbits(a, axis=1, count=6)
c

array([[0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 1],
       [0, 0, 0, 1, 0, 1]], dtype=uint8)

## 输出格式
`binary_repr(num, width=None)` 函数用于将输入数组的整数以二进制格式输出。


In [138]:
np.binary_repr(5)

'101'

In [139]:
np.binary_repr(-5)

'-101'

In [140]:
np.binary_repr(5, width=5)

'00101'

In [141]:
np.binary_repr(5, width=4)

'0101'